src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -84,23 +84,42 @@ private string GetCacheKeyPrivate(SecurityKey securityKey, string algorithm, str
             return string.Format(CultureInfo.InvariantCulture,
                                  "{0}-{1}-{2}-{3}",
                                  securityKey.GetType(),
-                                 string.IsNullOrEmpty(securityKey.KeyId) ? securityKey.InternalId : securityKey.KeyId,
+                                 securityKey.InternalId,
                                  algorithm,
                                  typeofProvider);
         }
 
+        /// <summary>
+        /// For some security key types, in some runtimes, it's not possible to extract public key material and create an <see cref="SecurityKey.InternalId"/>.
+        /// In these cases, <see cref="SecurityKey.InternalId"/> will be an empty string, and these keys should not be cached.
+        /// </summary>
+        /// <param name="signatureProvider"><see cref="SignatureProvider"/> to be examined.</param>
+        /// <returns><c>True</c> if <paramref name="signatureProvider"/> should be cached, <c>false</c> otherwise.</returns>
+        private bool ShouldCacheSignatureProvider(SignatureProvider signatureProvider)
+        {
+            if (signatureProvider.Key.InternalId.Length == 0)
```

Comment:
could we write: return !string.IsNullOrEmpty(signatureProvider.Key.InternalId);

Reply:
IsNullOrEmpty checks both for null and that length of the string is 0.
As InternalId can't be null we can skip the extra check.

I can remove the if-branching.

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1346/commits/303138374134fe6c096ded0ade0b448252f8f695.